hdk/assembly/emulator.py

```diff
@@ -0,0 +1,109 @@
+from array import array
+from collections.abc import Callable, Iterable
+
+from hdk.assembly.code import link_instructions
+from hdk.assembly.syntax import AInstruction, Command, Instruction
+
+_JUMP_TABLE: dict[str | None, Callable[[int], bool]] = {
+    None: lambda comp_result: False,
+    "JMP": lambda comp_result: True,
+    "JLE": lambda comp_result: comp_result <= 0,
+    "JNE": lambda comp_result: comp_result != 0,
+    "JLT": lambda comp_result: comp_result < 0,
+    "JGE": lambda comp_result: comp_result >= 0,
+    "JEQ": lambda comp_result: comp_result == 0,
+    "JGT": lambda comp_result: comp_result > 0,
+}
+
+
+_COMP_TABLE: dict[str, Callable[[int, int, int], int]] = {
+    "0": lambda a, d, m: 0,
+    "1": lambda a, d, m: 1,
+    "-1": lambda a, d, m: -1,
+    "D": lambda a, d, m: d,
+    "A": lambda a, d, m: a,
+    "!D": lambda a, d, m: ~d,
+    "!A": lambda a, d, m: ~a,
+    "-D": lambda a, d, m: -d,
+    "-A": lambda a, d, m: -a,
+    "D+1": lambda a, d, m: d + 1,
+    "A+1": lambda a, d, m: a + 1,
+    "D-1": lambda a, d, m: d - 1,
+    "A-1": lambda a, d, m: a - 1,
+    "D+A": lambda a, d, m: d + a,
+    "D-A": lambda a, d, m: d - a,
+    "A-D": lambda a, d, m: a - d,
+    "D&A": lambda a, d, m: d & a,
+    "D|A": lambda a, d, m: d | a,
+    "M": lambda a, d, m: m,
+    "!M": lambda a, d, m: ~m,
+    "-M": lambda a, d, m: -m,
+    "M+1": lambda a, d, m: m + 1,
+    "M-1": lambda a, d, m: m - 1,
+    "D+M": lambda a, d, m: d + m,
+    "D-M": lambda a, d, m: d - m,
+    "M-D": lambda a, d, m: m - d,
+    "D&M": lambda a, d, m: d & m,
+    "D|M": lambda a, d, m: d | m,
+}
+
+
+def run(instructions: Iterable[Instruction], steps: int, memory: array) -> None:
+    """Modifies memory by instructions
+
+    Args:
+        instructions: An iterable of symbolic instructions.
+        steps: number of commands to execute
+        memory: an array of signed int
+
+    Returns:
```

Comment:
Remove redundant "Returns" and empty lines. Each docstring section should be a valid English sentence starting with a capital and ending with a dot.